kasa/tests/common.py

```diff
@@ -0,0 +1,102 @@
+import glob
+import json
+import os
+from collections import namedtuple
+from os.path import basename
+from pathlib import Path
+from typing import List, Optional, Set
+
+FixtureInfo = namedtuple("FixtureInfo", "data protocol name")
+FixtureInfo.__hash__ = lambda x: hash((x.name, x.protocol))  # type: ignore[attr-defined, method-assign]
+FixtureInfo.__eq__ = lambda x, y: hash(x) == hash(y)  # type: ignore[method-assign]
+
+
+SUPPORTED_IOT_DEVICES = [
+    (device, "IOT")
+    for device in glob.glob(
+        os.path.dirname(os.path.abspath(__file__)) + "/fixtures/*.json"
+    )
+]
+
+SUPPORTED_SMART_DEVICES = [
+    (device, "SMART")
+    for device in glob.glob(
+        os.path.dirname(os.path.abspath(__file__)) + "/fixtures/smart/*.json"
+    )
+]
+
+
+SUPPORTED_DEVICES = SUPPORTED_IOT_DEVICES + SUPPORTED_SMART_DEVICES
+
+
+def idgenerator(paramtuple: FixtureInfo):
+    try:
+        return paramtuple.name + (
+            "" if paramtuple.protocol == "IOT" else "-" + paramtuple.protocol
+        )
+    except:  # TODO: HACK as idgenerator is now used by default  # noqa: E722
+        return None
+
+
+def get_fixture_info() -> List[FixtureInfo]:
+    """Return raw discovery file contents as JSON. Used for discovery tests."""
+    fixture_data = []
+    for file, protocol in SUPPORTED_DEVICES:
+        p = Path(file)
+        folder = Path(__file__).parent / "fixtures"
+        if protocol == "SMART":
+            folder = folder / "smart"
+        p = folder / file
+
+        with open(p) as f:
+            data = json.load(f)
+
+        fixture_name = basename(p)
+        fixture_data.append(
+            FixtureInfo(data=data, protocol=protocol, name=fixture_name)
+        )
+    return fixture_data
+
+
+FIXTURE_DATA: List[FixtureInfo] = get_fixture_info()
+
+
+def filter_fixtures(
+    desc,
+    *,
+    data_root_filter: Optional[str] = None,
+    protocol_filter: Optional[Set[str]] = None,
+    model_filter: Optional[Set[str]] = None,
+    component_filter: Optional[str] = None,
+):
+    filtered = []
```

Comment:
Please add a docstring explaining the different filters, it doesn't have to be comprehensive but just to remind how this is used and what how the filtering works on high-level. Your future self will thank you for doing that :-)

Reply:
Done